OpenTelemetry TraceIdRatioBased sampler requirements following OTEP 235 (#4166)
open-telemetry/opentelemetry-specification: specification/trace/tracestate-probability-sampling.md

### Consistent Sampling Decision

A consistent sampling decision means that a positive sampling decision made for a particular span with probability p1 necessarily implies a positive sampling decision for any span belonging to the same trace if it is made with probability p2 >= p1.

### Rejection Threshold (T)

This is a 56-bit value directly derived from the sampling probability. One way to think about this is that this is the number of spans that would be *dropped* out of 2^56 considered spans. This is an alternative to the `p` value in the previous specification. The `p` value is limited to powers of two, while this supports a large range of values.

You can derive the rejection threshold from the sampling probability as follows:

```
Rejection_Threshold = (1 - Sampling_Probability) * 2^56.
```

For example, if the sampling probability is 100% (keep all spans), the rejection threshold is 0.

> Similarly, if the sampling probability is 1% (drop 99% of spans), the rejection threshold with 5 digits of precision would be (1-0.01) * 2^56 = 4458562600304640 = 0xfd70a00000000.

Author: Thank you!

The following examples are in Golang and Python3. They are intended as examples only for clarity, and not as a suggested implementation.

### Converting floating-point probability to threshold value

Threshold values are encoded with trailing zeros removed, which allows for variable precision. This can be accomplished by rounding, and there are several practical ways to do this with built-in string formatting libraries.

With up to 56 bits of precision available, implementations that use built-in floating point number support will be limited by the precision of the underlying number support.  One way to encode thresholds uses the IEEE 754-2008-standard hexadecimal floating point representation as a simple solution.  For example, in Golang,

```go
// ProbabilityToThresholdWithPrecision assumes the probability value is in the range
// [0x1p-56, 1] and precision is in the range [1, 14].
func ProbabilityToThresholdWithPrecision(probability float64, precision int) string {
    if probability == 1 {
        // Special case
        return "0"
    }
    // Raise precision by the number of leading 0s or Fs
    _, expF := math.Frexp(probability)

>     _, expR := math.Frexp(1 - probability)

Author: I would say it can be removed, yes. We believe fine-resolution probabilities close to 0 are important, but not close to 1. 

I put this in for symmetry--so that rounding behavior near 0 and 1 would be the same, but I can easily be convinced to remove this line. It would mean that probabilities near 1 round up to exactly 1, subject to precision.

>     _, expR := math.Frexp(1 - probability)

Reviewer: I'm in favor of removing it to keep the relative loss of precision consistent across multiples of 16; it feels somewhat arbitrary to increase the precision for values above `0.9375`. Increasing the precision by the number of leading 0s would change the first row to `0cccd` / `2.0077358064973794E-7`:

| Probability    | Threshold w/ precision=4 | abs(1 - AdjustedCount * Probability) |
|----------------|------------|--------------------------------------|
| 0.95 / 16 ** 0 | 0ccd       | 3.212386963991065E-6                 |
| 0.95 / 16 ** 1 | f0ccd      | 3.212386963991065E-6                 |
| 0.95 / 16 ** 6 | ffffff0ccd | 3.212386963991065E-6                 |
| 0.90 / 16 ** 0 | 199a       | 6.78173001933402E-6                  |
| 0.90 / 16 ** 1 | f199a      | 6.78173001933402E-6                  |
| 0.90 / 16 ** 6 | ffffff199a | 6.78173001933402E-6                  |

>     _, expR := math.Frexp(1 - probability)

Author: Thank you. I will raise this point in tomorrow's Sampling SIG.

>     _, expR := math.Frexp(1 - probability)

Author: Sampling SIG agreed on this point. I will update the PR to drop the indicated line of code and clarify that probabilities close to 1 round to exactly 1, that in this case precision counts leading 0s.

>     _, expR := math.Frexp(1 - probability)

Author: [44c8190](https://github.com/open-telemetry/opentelemetry-specification/pull/4166/commits/44c8190e4a4665d43d01ba13e119a8216920e4eb)

<!-- tocstop -->

</details>

## Overview

Sampling is an important lever to reduce the costs associated with collecting and processing telemetry data. It enables you to choose a representative set of items from an overall population.

There are two key aspects for sampling of tracing data. The first is that sampling decisions can be made independently for *each* span in a trace. The second is that sampling decisions can be made at multiple points in the telemetry pipeline. For example, the sampling decision for a span at span creation time could have been to **keep** that span, while the downstream sampling decision for the *same* span at a later stage (say in an external process in the data collection pipeline) could be to **drop** it.

For each of the above aspects, if we don't make **consistent** sampling decisions, we will end up with traces that are unusable and do not contain a coherent set of spans, because of the independent sampling decisions. Instead, we want sampling decisions to be made in a **consistent** manner so that we can effectively reason about a trace.

> This specification describes a mechanism to achieve such consistent sampling decisions using a mechanism called **Consistent Probability Sampling**. To achieve this, it uses two key building blocks. The first is a common source of randomness (R) that is available to all participants. This can be either a custom value (called `rv`) or taken from the trailing 7 bytes of the TraceID. The second is a concept of a rejection threshold (T). This is derived directly from a participant's sampling rate. This proposal describes how these two values should be propagated and how participants should use them to make sampling decisions.

Reviewer 2: Participants isn't defined, or part of the glossary. Perhaps something like this could help:

> common source of randomness (R), typically set by the sampling configuration at the root span and potentially changed down the pipeline at a Collector

Reviewer 2: The usage of "participant" here is also unclear:

> This is derived directly from a participant's sampling rate

Author: Now defined "participants, which includes a set of tracers and collectors".

> For more details about this specification, see [OTEP 235](https://github.com/open-telemetry/oteps/blob/main/text/trace/0235-sampling-threshold-in-trace-state.md).

Reviewer 3: Should we remove the link to OTEP 235 here (if it is going to be merged to the main spec)?

Author: I believe link updates will happen later, after the move.

## Definitions

### Sampling Probability

> Sampling probability is the likelihood that a span will be *kept*. Each participant can choose a different sampling probability for each span. For example, if the sampling probability is 0.25, around 25% of the spans will be kept.

Reviewer 2: I'm starting to think that you mean "tracer" here instead of "participant", potentially being "collector" when this is made not by a tracer. So, participant is "a tracer or a Collector" ?

> Sampling probability is valid in the range 2^-56 through 1.  Note that the zero value is not defined and that "never" sampling is not a form of probability sampling.

Reviewer 2: 2^-56 might seem a bit random for the non-initiated: would it be worth saying that this so that we have 7 bytes, matching the 7 bytes we get from the "randomness" (typically the 7 rightmost bytes from the trace ID)?

### Consistent Sampling Decision

A consistent sampling decision means that a positive sampling decision made for a particular span with probability p1 necessarily implies a positive sampling decision for any span belonging to the same trace if it is made with probability p2 >= p1.

### Rejection Threshold (T)

This is a 56-bit value directly derived from the sampling probability. One way to think about this is that this is the number of spans that would be *dropped* out of 2^56 considered spans. This is an alternative to the `p` value in the previous specification. The `p` value is limited to powers of two, while this supports a large range of values.

You can derive the rejection threshold from the sampling probability as follows:

```
Rejection_Threshold = (1 - Sampling_Probability) * 2^56.
```

For example, if the sampling probability is 100% (keep all spans), the rejection threshold is 0.

Similarly, if the sampling probability is 1% (drop 99% of spans), the rejection threshold with 5 digits of precision would be (1-0.01) * 2^56 = 4458562600304640 = 0xfd70a00000000.

> We refer to this rejection threshold conceptually as `T`. We represent it using the key `th`. This must be propagated in both the `tracestate` header and in the TraceState attribute of each span.

Reviewer 2: ```suggestion
We refer to this rejection threshold conceptually as `T`. We represent it using the key `th`. This must be propagated in both the `tracestate` header and in the TraceState attribute of each span. In the example above, the `th` key has `fd70a00000000` as the value.
```

> We refer to this rejection threshold conceptually as `T`. We represent it using the key `th`. This must be propagated in both the `tracestate` header and in the TraceState attribute of each span. In the example above, the `th` key has `fd70a00000000` as the value.

Author: This reads as non-normative text to me. (The document isn't very strict about separating normative and non-normative sections, the way the previous document was. Not sure the rigor is worth it!) I changed the sentence to:

```
We refer to this rejection threshold conceptually as `T`. We represent it using the OpenTelemetry TraceState key `th`, where the value is propagated and also stored with each span. In the example above, the `th` key has `fd70a00000000` as the value.
```

See [tracestate handling](./tracestate-handling.md#sampling-threshold-value-th) for details about encoding threshold values.

### Random Value (R)

A common random value (that is known or propagated to all participants) is the main ingredient that enables consistent probability sampling. Each participant can compare this value (R) with their rejection threshold (T) to make a consistent sampling decision across an entire trace (or even across a group of traces).

This proposal supports two sources of randomness:

> - **A custom source of randomness**: This proposal allows for a *random* (or pseudo-random) 56-bit value. We refer to this as `rv`. This can be generated and propagated through the `tracestate` header and the tracestate attribute in each span.

Reviewer 2: I think I commented this elsewhere, but when should I, as a user, should consider having a custom source of randomness?

Author: This is meant to be part of #4162 which focuses on randomness. It writes "To enable sampling in this and other situations where TraceIDs lack sufficient randomness,"

However, I tried to stay away from the advanced use-cases some might mention. If you have a reason to use independent trace IDs and still want them to sample consistently, this is what you'd choose.

Author: Continuing the connection with #4162. See this commit, which hopefully answers this question ([thread](https://github.com/open-telemetry/opentelemetry-specification/pull/4162#discussion_r1766728589)).

- **Using TraceID as a source of randomness**: This proposal introduces using the last 56 bits of the `traceid` as the source of randomness. This can be done if the root participant knows that the `traceid` has been generated in a random or pseudo-random manner.

See [tracestate handling](./tracestate-handling.md#sampling-randomness-value-rv) for details about encoding randomness values.

### Consistent Sampling Decision Approach

Given the above building blocks, let's look at how a participant can make consistent sampling decisions. For this, two values MUST be present in the `SpanContext`:

1. The common source of randomness: the 56-bit `R` value.
2. The rejection threshold: the 56-bit `T` value.

If `R` >= `T`, *keep* the span, else *drop* the span.

> `T` represents the maximum threshold that was applied in all previous consistent sampling stages. If the current sampling stage applies a greater threshold value than any stage before, it MUST update (increase) the threshold correspondingly.

Reviewer 2: Perhaps this comes later, but the OTEP also mentions that this cannot be lowered, only increased.

Reviewer 2: I just came to the part where it says that it can be lowered at head samplers, but not for downstream samplers. This statement here might need to be adjusted then.

Author: Probability can be lowered after the fact by re-sampling with a higher threshold, but not raised after the fact (by a lower threshold). I'll look for any inconsistencies.

First, a consistent probability `Sampler` may choose its own sampling rate. The higher the chosen sampling rate, the lower the rejection threshold (T). It MAY select any value of T. If a valid `SpanContext` is provided in the call to `ShouldSample` (indicating that the span being created will be a child span), there are two possibilities:

- **The child span chooses a T greater than the parent span's T**: The parent span may be *kept* but it is possible that its child, the current span, may be dropped because of the lower sampling rate. At the same time, in the case where the decision for the child span is to *keep* it, the decision for the parent span would have also been to *keep* (due to our consistent sampling approach) since the parent's sampling rate is greater than the child's sampling rate.

> - **The child span chooses a T less than or equal to the parent span's T**:  The parent span might have been *dropped* but it is possible that its child, the current span, may be *kept* because of the higher sampling rate. At the same time, in case where the parent span is *kept*, the child span would be *kept* as well (due to our consistent sampling approach) since the child's sampling rate is greater than the parent's sampling rate.

Reviewer 4: The phrase "parent span might have been dropped" is not correct in this context. According to our specs, if the parent span was dropped, there wouldn't be parent's T-value available (we clear `th` value when dropping a span. I suggest saying "Downstream samplers may decide to drop the parent span".

Author: I read this section and it has me a bit confused. I propose to cut a lot of text, since I'm not sure it was answering real questions.

Note that while both the above cases can result in incomplete traces, they still meet the consistent sampling goals.

For the output TraceState,

- The `th` key MUST be defined with a value corresponding to the sampling probability the sampler used.
- The `rv` value, if present on the input TraceState, MUST be defined and equal to the incoming span context's `rv` value, including the root context.

> Trace SDKs are responsible for for synthesizing `rv` values in the OpenTelemetry TraceState root span contexts.

Reviewer 2: ```suggestion
Trace SDKs are responsible for synthesizing `rv` values in the OpenTelemetry TraceState root span contexts.
```

### Downstream samplers

A downstream sampler, in contrast, may output a given ended Span with a *modified* trace state, complying with following rules:

- If the chosen sampling probability is 1, the sampler MUST NOT modify any existing `th`, nor set any `th`.

> - Otherwise, the chosen sampling probability is in `(0, 1)`. In this case the sampler MUST output the span with a `th` equal to `max(input th, chosen th)`. In other words, `th` MUST NOT be decreased (as it is not possible to retroactively adjust an earlier stage's sampling probability), and it MUST be increased if a lower sampling probability was used. This case represents the common case where a downstream sampler is reducing span throughput in the system.

Reviewer 2: ```suggestion
- Otherwise, the chosen sampling probability is in `[0, 1)`. In this case the sampler MUST output the span with a `th` equal to `max(input th, chosen th)`. In other words, `th` MUST NOT be decreased (as it is not possible to retroactively adjust an earlier stage's sampling probability), and it MUST be increased if a lower sampling probability was used. This case represents the common case where a downstream sampler is reducing span throughput in the system.
```

Author: We can't represent 0-probability sampling with a threshold. The corresponding threshold is out-of-range on purpose. If you want 0-probability sampling, you can simply not export a span, or you can export a span w/o the `th` value set, which says "unknown sampling probability".

### Migration to consistent probability samplers

The OpenTelemetry specification for TraceIdRatioBased samplers was not completed until after the SDK specification was declared stable, and the exact behavior of that sampler was left unspecified.  The `th` and `rv` sub-keys defined in the OpenTelemetry TraceState now address this behavior specifically.

As the OpenTelemetry TraceIdRatioBased sampler changes definition, users must consider how to avoid incomplete traces due to inconsistent sampling during the transition between old and new logic.

The original TraceIdRatioBased sampler specification gave a workaround for the underspecified behavior, that it was safe to use for root spans: "It is recommended to use this sampler algorithm only for root spans (in combination with [`ParentBased`](./sdk.md#parentbased)) because different language SDKs or even different versions of the same language SDKs may produce inconsistent results for the same input."

> To avoid inconsistency during this transition, users SHOULD follow this guidance until all TraceIdRatioBased samplers used in a system have been upgraded to the modern `TraceIdRatioBased` specification based on W3C Trace Context Level 2 randomness.  After all `TraceIdRatioBased` samplers have been upgraded, it is safe to use `TraceIdRatioBased` sampler without also using the `ParentBased` sampler.

Reviewer 2: How can users assess that they reached this? Should we keep a table, showing from which versions which SDKs support the new spec?

Author: Another way they can do this is to wait for all spans to have the W3C trace Random flag set across a system. How does that sound?

Author: I came back to this point and wrote up a stronger answer to your question.

First answer: users can know when all head samplers support W3C Trace Context Level 2 when all spans they collect have the Trace random flag set in their Span flags.

I think this is a legitimate concern, and I propose specifying a compatibility warning statement that SDKs can issue to warn users that their system is not completely upgraded. The warning would be issued once per SDK lifetime, to avoid noise, and tell the user that the TraceIdRatioBased sampler made a decision that presumed TraceID randomness but did not have the random flag set. This way, after a user believes they've fully migrated, they will still receive a warning to indicate a problem.

Author: See 
1855839

## Algorithms

The `th` and `rv` values may be represented and manipulated in a variety of forms depending on the capabilities of the processor and needs of the implementation. As 56-bit values, they are compatible with byte arrays and 64-bit integers, and can also be manipulated with 64-bit floating point with a truly negligible loss of precision.

The following examples are in Golang and Python3. They are intended as examples only for clarity, and not as a suggested implementation.

### Converting floating-point probability to threshold value

> Threshold values are encoded with trailing zeros removed, which allows for variable precision.  This can be accompolished by rounding, and there are several practical way to do this with built-in string formatting libraries.

Reviewer 2: ```suggestion
Threshold values are encoded with trailing zeros removed, which allows for variable precision. This can be accomplished by rounding, and there are several practical ways to do this with built-in string formatting libraries.
```

> With up to 56 bits of precision available, implementations that use built-in floating point number support will be limited by the precision of the underlying number support.  If the language supports IEEE 754-2008-standard hexadecimal floating point, for example in Golang,

Reviewer 2: The last statement sounds a bit strange.

Author: Yes. Fixed:

```
One way to encode thresholds uses the IEEE 754-2008-standard hexadecimal floating point representation as a simple solution.  For example, in Golang,
```

> With up to 56 bits of precision available, implementations that use built-in floating point number support will be limited by the precision of the underlying number support.  One way to encode thresholds uses the IEEE 754-2008-standard hexadecimal floating point representation as a simple solution.  For example, in Golang,

Reviewer 5: works out naturally.

    if probability == 1 {
        // Special case
        return "0"
    }
    // Raise precision by the number of leading 0s or Fs
    _, expF := math.Frexp(probability)
    _, expR := math.Frexp(1 - probability)
    precision = min(14, max(precision+expF/-4, precision+expR/-4))

    // Change the probability to rejection probability, with range [0, 1),
    // translate rejection probability by +1, into range [1, 2).
    // Format the significand of this expression as hexadecimal floating point
    asHex := strconv.FormatFloat(2-probability, 'x', precision, 64)

    // Strip the leading "0x1.", use the requested number of digits.
    // Strip additional trailing zeros.
    digits := asHex[4 : 4+precision]
    return strings.TrimRight(digits, "0")
}
```

To translate directly from floating point probability into a 56-bit unsigned integer representation using `math.Round()` and shift operations, see the [OpenTelemetry Collector-Contrib `pkg/sampling` package][PKGSAMPLING] package demonstrates this form of directly calculating integer thresholds from probabilities.

[PKGSAMPLING]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/sampling/README.md

> OpenTelemetry SDKs are recommended to use 4 digits of precision by default. The following table shows values computed by the method above for 1-in-N probability sampling, with precision 3, 4, and 5.

Reviewer 6: Why 4?

Author: This is a little bit arbitrary. When I updated the OTel-Collector's `processor/probabilisticsamplerprocessor` I found an existing hash function that used 14 bits of information; in that case, 4 hex-digits of precision was enough to convey that sampler's decision w/o loss.

Compare and contrast: the TraceIdRatioBased sampler in the existing specification doesn't say what form of floating point manipulation is used to get to a ratio, though it recommends 6 decimal places of precision when printing that number in the description.

We could analyze the relative error to make this decision (@oertl what do you think?), however there's an argument that it matters less than you might think. When we interpret a threshold value, it is an exact representation of the sampling process. Sampling probabilities represented by thresholds are always exact--when you see a span with the threshold you have a corresponding, exact rational number telling you its probability (i.e., (2^56 - T)/(2^56)).

The question of accuracy and precision here applies to the translation from an input representation (likely a percentage, OTel doesn't specify this) to the exact threshold that will be used. If users are turning sampling probabilities into integer adjusted counts, this precision determines the error that will be introduced. I think 3 digits yields too much error (e.g., 1/100 sampling w/ adjusted counts of 100.05 is a 0.05% error) and 4 digits is acceptable (e.g., 1/10 sampling w/ adjusted counts of 99.9997 is a 0.003% error). @tsloughter what do you think?

> <!--- Program at https://go.dev/play/p/7eLM6FkuoA5 (includes function above) generates the table below --->

Reviewer 7: I authentically love this table, but I'd also love to see two things changed about it:
1) Invert the order (1,2,3 first) -- easier to understand that way.
2) Never use scientific notation. It just makes the table harder to read. (And I don't know if it's possible in markdown, but it would be awesome if the decimal points could align.)

Author: Fixed!

| 1000    | 0.001              | ffbe7<br/>ffbe77<br/>ffbe76d       | 0.0010004043579101562<br/>0.0009999871253967285<br/>0.000999998301267624     | 999.5958055290753<br/>1000.012874769029<br/>1000.0016987352618        |
| 10000   | 0.0001             | fff972<br/>fff9724<br/>fff97247    | 0.00010001659393310547<br/>0.00010000169277191162<br/>0.00010000006295740604 | 9998.340882002383<br/>9999.830725674266<br/>9999.99370426336          |
| 100000  | 0.00001              | ffff584<br/>ffff583a<br/>ffff583a5 | 9.998679161071777e-06<br/>1.00000761449337e-05<br/>1.0000003385357559e-05    | 100013.21013412817<br/>99999.238556461<br/>99999.96614643588          |
| 1000000 | 0.000001              | ffffef4<br/>ffffef39<br/>ffffef391 | 9.98377799987793e-07<br/>1.00000761449337e-06<br/>9.999930625781417e-07      | 1.0016248358208955e+06<br/>999992.38556461<br/>1.0000069374699865e+06 |

### Converting integer threshold to a `th`-value

To convert a 56-bit integer threshold value to the t-value representation, emit it as a hexadecimal value (without a leading '0x'), optionally with trailing zeros omitted:

```py
h = hex(tvalue).rstrip('0')
# remove leading 0x
tv = 'tv='+h[2:]

> ```

Reviewer 4: Nit: wouldn't this code result in an empty string if the threshold is zero?

Author: Yes. I propose

```
if tvalue == 0: 
  add_otel_trace_state('tv:0')
else:
  h = hex(tvalue).rstrip('0')
  # remove leading 0x
  add_otel_trace_state('tv:'+h[2:])
```

Reviewer 4: Sounds good

<!-- toc -->

- [Overview](#overview)

> - [Definitions](#definitions)

Author: This diff is nonsensical. If we could merge https://github.com/open-telemetry/opentelemetry-specification/pull/4168 before this PR, the diff would look empty because this file contains entirely new content.
